plugin/storage/badger/spanstore/reader.go

```diff
@@ -275,81 +274,138 @@ func serviceQueries(query *spanstore.TraceQueryParameters, indexSeeks [][]byte)
 		prevTraceID := []byte{}
-		ids = append(ids, make([][]byte, 0, len(indexResults)))
-		for _, k := range indexResults {
-			traceID := k[len(k)-sizeOfTraceID:]
+		innerIDs := make([][]byte, 0, len(indexSeeks))
+		for j := 0; j < len(indexResults); j++ {
+			traceID := indexResults[j]
 			if !bytes.Equal(prevTraceID, traceID) {
-				ids[i] = append(ids[i], traceID)
+				innerIDs = append(innerIDs, traceID)
 				prevTraceID = traceID
 			}
 		}
-		sort.Slice(ids[i], func(k, h int) bool {
-			return bytes.Compare(ids[i][k], ids[i][h]) < 0
-		})
+
+		// Merge-join current results
+		if plan.mergeOuter == nil {
+			plan.mergeOuter = innerIDs
+		} else {
+			plan.mergeOuter = mergeJoinIds(plan.mergeOuter, innerIDs)
+		}
+	}
+
+	// Last scan should get us in correct timestamp order
+	ids, err := r.scanIndexKeys(indexSeeks[0], plan)
+	if err != nil {
+		return nil, err
+	}
+
+	if plan.mergeOuter != nil {
+		// Build hash of the current merged data
+		plan.hashOuter = buildHash(plan, plan.mergeOuter)
+		plan.mergeOuter = nil
+	} else {
+		// We filter the last elements
+		plan.hashOuter = buildHash(plan, ids)
+	}
+
+	traceIDs := filterIDs(plan, ids)
+	return traceIDs, nil
+}
+
+func filterIDs(plan *executionPlan, innerIDs [][]byte) []model.TraceID {
+	traces := make([]model.TraceID, 0, plan.limit)
+
+	items := 0
+	for i := 0; i < len(innerIDs); i++ {
+		trID := bytesToTraceID(innerIDs[i])
+
+		if _, found := plan.hashOuter[trID]; found {
+			traces = append(traces, trID)
+			delete(plan.hashOuter, trID) // Prevent duplicate add
+			items++
+		}
+
+		if items == plan.limit {
+			return traces
+		}
+	}
+
+	return traces
+}
+
+func bytesToTraceID(key []byte) model.TraceID {
+	return model.TraceID{
+		High: binary.BigEndian.Uint64(key[:8]),
+		Low:  binary.BigEndian.Uint64(key[8:sizeOfTraceID]),
 	}
-	return ids, nil
 }
 
-// durationQueries checks non unique index of durations
-func (r *TraceReader) durationQueries(query *spanstore.TraceQueryParameters, ids [][][]byte) [][][]byte {
+func buildHash(plan *executionPlan, outerIDs [][]byte) map[model.TraceID]struct{} {
+	var empty struct{}
+
+	hashed := make(map[model.TraceID]struct{})
+	for i := 0; i < len(outerIDs); i++ {
+		trID := bytesToTraceID(outerIDs[i])
+
+		if plan.hashOuter != nil {
+			if _, exists := plan.hashOuter[trID]; exists {
+				hashed[trID] = empty
+				delete(plan.hashOuter, trID) // Filter duplications
+			}
+		} else {
+			hashed[trID] = empty
+		}
+	}
+
+	return hashed
+}
+
+// durationQueries checks non unique index of durations and returns a map for further filtering purposes
+func (r *TraceReader) durationQueries(plan *executionPlan, query *spanstore.TraceQueryParameters) map[model.TraceID]struct{} {
 	durMax := uint64(model.DurationAsMicroseconds(query.DurationMax))
 	durMin := uint64(model.DurationAsMicroseconds(query.DurationMin))
 
-	startKey := make([]byte, 0, 9)
-	endKey := make([]byte, 0, 9)
+	startKey := make([]byte, 9)
```

Comment:
9 == 1 + 8, right? In other places in this source, you are using 1+8, which I find clearer than simply 9.

Reply:
Yes, it's the length of timestamp + indexKey (8+1).